[FEAT] Common class: InputTextView

soobin-k opened the pull request:
## 📌 PR 요약

🌱 작업한 내용
- 텍스트 입력을 받아오는 UI가 반복되는 것 같아 common class에서 불러와 쓸 수 있게 구현해보았습니다.

🌱 PR 포인트
- 제목, 텍스트뷰의 디자인은 고정이고
- 글자수 세기 라벨, 물음표 버튼은 때에 따라 다르므로 옵션 값으로 지정이 가능합니다.

## 📸 스크린샷

|스크린샷|
|:--:|
|<img width="398" alt="스크린샷 2023-01-09 오전 12 24 18" src="https://user-images.githubusercontent.com/77331348/211204705-a9b76456-c21e-4235-8ead-9fd8e12d4094.png">|
|<img width="402" alt="스크린샷 2023-01-09 오전 12 25 55" src="https://user-images.githubusercontent.com/77331348/211204811-2ba84924-9eda-4d9d-be9e-55ac67cc6138.png">
|<img width="484" alt="스크린샷 2023-01-09 오전 12 26 41" src="https://user-images.githubusercontent.com/77331348/211204855-626b3df9-5dff-4290-86f0-3c9aa8cf17dd.png">

## 테스트 코드
- 옵션값 지정
    
    ```jsx
  private let introduceTitleView = InputTextView(
      title: "소개 타이틀",
      placeHolder: "날 좋은날 같이 사진 찍으러 갈 사람~",
      options: [.textCount, .questionMark]
  )
    ```
    
- 옵션값 생략
    
    ```jsx
  private let nameTitleView = InputTextView(
      title: "모임 이름",
      placeHolder: "우리동네 사진모임"
  )
    ```
## 📮 관련 이슈

soobin-k committed:
[ADD] Folder & ViewController

soobin-k committed:
[ADD] questionButton 이미지 추가

soobin-k committed:
[ADD] InputTextView 추가

soobin-k committed:
[DEL] InputTextView placeholder 제거

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,131 @@
+  
+  let textView = UITextView().then {
+    $0.textColor = .deepGray
+    $0.textContainerInset = UIEdgeInsets(
+      top: 14,
+      left: 8,
+      bottom: 14,
+      right: 8
+    )
```
top과 bottom은 figma상으로 11만큼 떨어져있는데 14로 정해두신 이유가 궁금해요!

soobin-k replied:
헛 제가 잘못봤나보네요! 바로 수정해놓겠습니다!

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,131 @@
+      countTextLabel.snp.makeConstraints {
+        $0.height.equalTo(12)
+      }
+      stackView.setCustomSpacing(4, after: textView)
```
좋은 메서드 알려주셔서 감사합니다 ☺️ 처음알았어요

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,131 @@
+    $0.setImage(UIImage(named: "questionButton"), for: .normal)
+  }
+  
+  let textView = UITextView().then {
+    $0.textColor = .deepGray
+    $0.textContainerInset = UIEdgeInsets(
+      top: 14,
+      left: 8,
+      bottom: 14,
+      right: 8
+    )
+    $0.font = .body2
+    $0.backgroundColor = .white
+    $0.layer.cornerRadius = 8
+    $0.isScrollEnabled = false
+  }
+  
+  let countTextLabel = UILabel().then {
+    $0.font = .overLine
+    $0.textAlignment = .right
+  }
```
일부로 `private`을 선언하지 않으신 걸까요?

soobin-k replied:
- textView: VC에서 rx나 delegate 사용하기 위해 private으로 선언하지 않았습니다!
- countTextLabel text에 직접 값을 지정해야할 경우가 생겨서 private 선언하지 않았습니다.
-> 이부분도 수정해보겠습니다!

dlrjswns replied:
개인적으로 프로퍼티는 의도치않는 수정이 있을 수 있으니 private으로 선언하고 rx혹은 delegate는 InputTextView만의 rx delegate proxy를 새로 만들어서 반영해보면 어떨까요 👀👀 ??

soobin-k replied:
우선 rx를 활용하여 모두 private으로 변경해보았습니다!
추후 rx delegate proxy를 활용해 디벨롭 해보겠습니다!😊✨

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,131 @@
+    $0.textAlignment = .right
+  }
+  
+  init(
+    title: String,
+    placeHolder: String,
+    option: InputTextOption = InputTextOption(
+      textCount: false,
+      questionOption: false
+      )
+  ) {
```
만약 textCount만을 필요로 했을 때 questionOption도 명시적으로 `false`를 처리해야하는데, OptionSet을 이용해보는 건 어떨까요?
```swift
struct InputViewOptions: OptionSet {
  let rawValue: UInt
  
  static let textCount = InputViewOptions(rawValue: 1 << 0)
  static let questionMark = InputViewOptions(rawValue: 1 << 1)
}
```

```swift
 init(
    title: String,
    placeHolder: String,
    options: InputViewOptions = []
)
```
위와 같이 이용하면 나중에 부를 때 좀 더 편리하게 사용될 수 있을 것 같아서요 :)

```swift
InputTextView(
    title: "소개 타이틀",
    placeHolder: "날 좋은날 같이 사진 찍으러 갈 사람~",
    options: .textCount
)

InputTextView(
    title: "소개 타이틀",
    placeHolder: "날 좋은날 같이 사진 찍으러 갈 사람~",
    options: .questionMark
)

InputTextView(
    title: "소개 타이틀",
    placeHolder: "날 좋은날 같이 사진 찍으러 갈 사람~",
    options: [.textCount, .questionMark]
)
```

soobin-k replied:
너무 좋은 것 같습니다! 알려주셔서 감사해요~! 
바로 반영해놓겠습니다!😊

dlrjswns replied:
OptionSet이라는것도 있군요 승현님 최고입니다 👍

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,131 @@
+    option: InputTextOption = InputTextOption(
+      textCount: false,
+      questionOption: false
+      )
```
아 그리구 여기는 띄어쓰기가 2칸 더 들어가있는 것 같아요.

```diff
    option: InputTextOption = InputTextOption(
      textCount: false,
      questionOption: false
-     )
+   )
```

soobin-k replied:
바로 수정해놓겠습니다..! 세세한 리뷰 감사합니다🙏🏻👍

WhiteHyun reviewed (commented):
1. 사용자가 textView를 누를 시 placeholder 제거
2. 입력없이 `resign first responder` 시 텍스트가 없을 때 placeholder 재생성
3. count label의 업데이트

위 기능은 나중에 구현 예정일까요?

soobin-k committed:
[CHORE] textView top, bottom inset 변경

soobin-k commented:
> 

이부분도 같이 구현해보겠습니다!

soobin-k committed:
[FEAT] InputTextView 기능 추가 with RxSwift
- 사용자가 textView를 누를 시 placeholder 제거
- 입력없이 resign first responder 시 텍스트가 없을 때 placeholder 재생성
- count label의 업데이트

soobin-k committed:
[REFACTOR] InputViewOptions를 OptionSet 활용하도록 변경

soobin-k committed:
[FEAT] 모임 생성 > 모임 이름 설정뷰 일부 구현
- InputTextView 테스트 코드 추가

soobin-k commented:
리뷰해주신 부분 반영완료했습니다!!😊
또한 MeetingNameViewController에 간단하게 적용해보았습니다. 참고 부탁드립니다!🙏🏻✨

https://user-images.githubusercontent.com/77331348/211330002-998f3adc-422b-4f6d-a76e-53d5bdc723ce.mov



@WhiteHyun @dlrjswns

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,227 @@
+//  Created by 김수빈 on 2023/01/08.
+//
+
+import SnapKit
+import UIKit
+import RxSwift
```
내장 프레임워크를 먼저 임포트하고, 빈 줄로 구분하여 서드파티 프레임워크를 임포트해주실 수 있나요?

```suggestion
import UIKit

import RxSwift
import SnapKit
```

dlrjswns replied:
저도 이 부분 수정해서 작성해야겠네요 !! 승현님 감사합니다 😁😁

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,227 @@
+    updateWrittenCharactersLabel(count: 0, pointColor: .mediumGray)
+  }
+  
+  private func bind(){
```
```suggestion
  private func bind() {
```

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,227 @@
+    textView.rx.didBeginEditing
+      .subscribe(onNext: { [weak self] in
+        guard let self = self else { return }
+        if(self.textView.text == self.placeHolder &&
+           self.textView.textColor == .deepGray){
```
띄어쓰기와 소괄호 제거 가능한가요?
만약 소괄호를 사용한 이유가 가독성 때문이라면, 제거 안하셔도 상관 없으나 나머지 밑에도 소괄호를 붙여서 일관된 코드 작성을 유지하는 게 좋을 것 같아요. 😄

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,227 @@
+    textView.rx.didEndEditing
+      .subscribe(onNext: { [weak self] in
+        guard let self = self else { return }
+        if self.textView.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
+            self.textView.textColor == .black {
```
두 번째 `self`가 한 칸 띄워져있네요.

dlrjswns replied:
저희 이 부분 이전 회의 코드컨벤션이야기할때처럼 self 지울 수 있다면 지우는것도 고려해보면 좋을꺼같아요 !!👍👍

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,227 @@
+  
+    textView.rx.text
+      .asDriver()
+      .drive (onNext: { [weak self] in
```
drive 뒤에 열린 소괄호 붙여주세요!

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,227 @@
+        guard let self = self else { return }
+        self.updateTextViewHeightAutomatically()
+        
+        guard $0 != self.placeHolder, // placeHolder
+                let text = $0 else { return }
```
Re-Indent 처리해주실 수 있나요?

WhiteHyun commented on PLUB/Sources/Views/Meeting/CreateMeeting/MeetingNameViewController.swift:
```diff
@@ -0,0 +1,83 @@
+//  Created by 김수빈 on 2023/01/08.
+//
+
+import UIKit
+import SnapKit
+import Then
```
이것도 맨 위와 같습니다.

soobin-k committed:
[REFACTOR] pr 리뷰 반영

soobin-k committed:
[REFACTOR] pr 리뷰 반영

soobin-k committed:
[REFACTOR] pr 리뷰 반영 - self 지우기

soobin-k commented:
리뷰 달아주신 부분 모두 반영 완료했습니다! 세세한 리뷰 감사드립니다!😊🙏🏻
@dlrjswns @WhiteHyun

WhiteHyun commented on PLUB/Configuration/CommonClass/InputTextView.swift:
```diff
@@ -0,0 +1,227 @@
+//  Created by 김수빈 on 2023/01/08.
+//
+
+import UIKit
+import RxSwift
+
+import SnapKit
```
RxSwift도 써드파티라서 다음과 같이 적용하시면 됩니다!
```suggestion
import UIKit

import RxSwift
import SnapKit
```

soobin-k replied:
앗 제가 제대로 반영을 안했었네요! 바로 수정해서 올리겠습니다!

WhiteHyun commented on PLUB/Sources/Views/Meeting/CreateMeeting/MeetingNameViewController.swift:
```diff
@@ -0,0 +1,84 @@
+//  Created by 김수빈 on 2023/01/08.
+//
+
+import UIKit
+import RxSwift
+
+import SnapKit
```
위와 동일합니다.

soobin-k committed:
[REFACTOR] pr 리뷰 반영 - 써드파티, 퍼스트파티 구분

WhiteHyun reviewed (approved):
굳! 고생하셨습니다.

soobin-k committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	PLUB.xcodeproj/project.pbxproj
